File: src/c#/lists/LinkedList.cs
using System;

namespace lists
{
  public class LinkedList<T>
  {
    private class Node
    {
      public T data;
      public Node next = null;

      public Node(T data)
      {
        this.data = data;
      }
    }

    private int size = 0;
    private Node _head = null;

    public int length() => size;
   
    public LinkedList<T> insert(T value)
    {
      size += 1;

      if (_head is null)
      {
        _head = new Node(value);
        return this;
      }

      Node current = _head;

      while (current.next != null)
      {
        current = current.next;
      }
      current.next = new Node(value);

      return this;
    }

    public LinkedList<T> remove_if(Func<T, bool> predicate)
    {
      if (predicate(_head.data))
      {
        _head = _head.next;
        size -= 1;
        return this;
      }

      Node current = _head;

      while (current.next != null)
      {
        if (predicate(current.data))
        {
          current.next = current.next.next;
          size -= 1;
          return this;
        }

        current = current.next;
      }

      return this;
    }

    public T find_if(Func<T, bool> predicate)
    {
      Node current = _head;

      while (current != null)
      {
        if (predicate(current.data))
        {
          return current.data;
        }

        current = current.next;
      }

      return default(T);
    }

    public bool includes(T value)
    {
      return find_if((v) => v.Equals(value)).Equals(value);
    }
  }

}

File: src/typescript/lists/linkedlist/LinkedList.ts
class Node<T> {
  public data: T;
  public next: Node<T>;

  constructor(data: T) {
    this.data = data;
  }
}

export default class LinkedList<T> {
  private _head: Node<T>;
  private _length: number = 0;

  public length = (): number => this._length;
  public head = (): Node<T> => this._head;

  private assert_index_is_valid = (index: number): void => {
    if (index < 0 || index > this._length)
      throw new Error(`Invalid index at Linkedlist.insert_at(${index}, T)`);
  };

  public insert = (value: T): LinkedList<T> =>
    this.insert_at(this._length, value);

  public insert_at = (index: number, value: T): LinkedList<T> => {
    this.assert_index_is_valid(index);

    this._length += 1;

    if (index == 0) {
      const temp: Node<T> = new Node(value);
      temp.next = this._head;
      this._head = temp;
      return this;
    }

    let current: Node<T> = this._head;
    let current_index: number = 0;
    while (current_index + 1 != index) {
      current = current.next;
      current_index += 1;
    }

    const temp: Node<T> = current.next;
    current.next = new Node(value);
    current.next.next = temp;

    return this;
  };

  public find_if = (predicate: (value: T) => boolean) => {
    let current: Node<T> = this._head;

    while (current) {
      if (predicate(current.data)) return current.data;

      current = current.next;
    }

    return null;
  };
}
